src/data/equiv/derangements/finite.lean

```diff
@@ -0,0 +1,117 @@
+/-
+Copyright (c) 2021 Henry Swanson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henry Swanson
+-/
+import data.equiv.basic
+import data.equiv.derangements.basic
+import data.fintype.basic
+import data.fintype.card
+import data.set.finite
+
+import tactic.ring
+import tactic.zify
+
+/-!
+# Derangements on fintypes
+
+This file contains lemmas that describe the cardinality of `derangements α` when `α` is a fintype.
+
+# Main definitions
+
+  - `num_derangements n`: The number of derangements on an n-element set. For concreteness, `fin n`
+    is used.
+
+  - `num_derangements_invariant`: A lemma stating that the number of derangements on a type `α`
+    depends only on the cardinality of `α`.
+
+  - `num_derangements_recursive`: A lemma giving a recursive expression for `num_derangements (n+2)`
+    in terms of `num_derangements n` and `num_derangements (n+1)`.
+
+  - `num_derangements_sum`: A lemma giving an expression for `num_derangements n` in terms of
+    factorials.
+-/
+
+open equiv fintype perm
+open_locale big_operators
+
+variables {α : Type*} [decidable_eq α] [fintype α]
+
+instance : decidable_pred (@is_derangement α) := λ _, fintype.decidable_forall_fintype
+
+instance : fintype (derangements α) :=
+begin
+  rw [derangements],
+  apply_instance
+end
+
+/-- The number of derangements on an `n`-element set. -/
+def num_derangements (n : ℕ) : ℕ := card (derangements (fin n))
+
+@[simp] lemma num_derangements_invariant (α : Type*) [fintype α] [decidable_eq α] :
+  card (derangements α) = num_derangements (card α) :=
+begin
+  apply card_eq.mpr,  -- card_eq because we don't need the specific equivalence
+  use derangements_congr (equiv_fin α),
+end
+
+theorem num_derangements_succ (n : ℕ) :
+  num_derangements (n+2) = (n+1) * num_derangements n + (n+1) * num_derangements (n+1) :=
+begin
+  have card_everything_but : ∀ a : fin (n+1), card ({a}ᶜ : set (fin (n+1))) = n,
+  { intro a,
+    simp only [fintype.card_of_finset, set.mem_compl_singleton_iff],
+    rw [finset.filter_ne' _ a, finset.card_erase_of_mem (finset.mem_univ a)],
+    simp },
+  have key := card_congr (@derangements_recursion_equiv (fin (n+1)) _),
+  rw [num_derangements_invariant, fintype.card_option, fintype.card_fin] at key,
+  simp [card_everything_but, mul_add, key]
+end
+
+lemma num_derangements_0 : num_derangements 0 = 1 := rfl
+
+lemma num_derangements_1 : num_derangements 1 = 0 := rfl
```

Comment:
This lemma and the one above should be `simp`

Reply:
Oddly enough, `simp [num_derangements]` times out for me here :(

Not entirely sure why simp doesn't know these are "defeq" (is that the term?)

Reply:
I think Eric meant `@[simp]`

Reply:
Oh! Yeah now I get it